internal/controllers/clusterextension_controller.go

```diff
@@ -557,6 +676,137 @@ func clusterExtensionRequestsForCatalog(c client.Reader, logger logr.Logger) han
 	}
 }
 
+type releaseState string
+
+const (
+	stateNeedsInstall releaseState = "NeedsInstall"
+	stateNeedsUpgrade releaseState = "NeedsUpgrade"
+	stateUnchanged    releaseState = "Unchanged"
+	stateError        releaseState = "Error"
+)
+
+func (r *ClusterExtensionReconciler) getReleaseState(cl helmclient.ActionInterface, obj metav1.Object, chrt *chart.Chart, values chartutil.Values, post *postrenderer) (*release.Release, releaseState, error) {
+	currentRelease, err := cl.Get(obj.GetName())
+	if err != nil && !errors.Is(err, driver.ErrReleaseNotFound) {
+		return nil, stateError, err
+	}
+	if errors.Is(err, driver.ErrReleaseNotFound) {
+		return nil, stateNeedsInstall, nil
+	}
+
+	desiredRelease, err := cl.Upgrade(obj.GetName(), r.ReleaseNamespace, chrt, values, func(upgrade *action.Upgrade) error {
+		upgrade.DryRun = true
+		return nil
+	}, helmclient.AppendUpgradePostRenderer(post))
+	if err != nil {
+		return currentRelease, stateError, err
+	}
+	if desiredRelease.Manifest != currentRelease.Manifest ||
+		currentRelease.Info.Status == release.StatusFailed ||
+		currentRelease.Info.Status == release.StatusSuperseded {
+		return currentRelease, stateNeedsUpgrade, nil
+	}
+	return currentRelease, stateUnchanged, nil
+}
+
+var GetInstalledbundle = func(ctx context.Context, acg helmclient.ActionClientGetter, allBundles []*catalogmetadata.Bundle, ext *ocv1alpha1.ClusterExtension) (*catalogmetadata.Bundle, error) {
```

Comment:
Why did this go from a method to a global variable (package scoped)? Doing this opens us up to bugs if this variable is somehow overwritten. I would highly recommend avoiding this.

Reply:
Instead of using a global variable, I would recommend creating a new interface like:

```go
type InstalledBundleGetter interface {
  GetInstalledBundle(...) (...)
}
```

and make use of it via a field in the reconciler like:
```go
type Reconciler struct {
  ...
  InstalledBundleGetter InstalledBundleGetter
  ...
}

func (r *Reconciler) reconcile(...) (...) {
  ...
  installedBundle, err := r.InstalledBundleGetter.GetInstalledBundle(...)
  ...
}
```

This has the ability to be mocked for testing while also preventing potential issues with using global variables

Reply:
Coming back around to this, something like:
```go
type Reconciler struct {
  GetInstalledBundleFunc func(...)(...)
}
```
_might_ be even better due to it being an explicit type that I don't believe would be allowed to be set as `nil`, alleviating the need to do a potential validation check for a nil reference

Reply:
Though not the best design for now, since it's internal and has no chances of external users breaking it - I'd suggest to do it in a follow up. Especially because it'll require a lot more bits of controller to be changed. An even better way would be for us to instead modify the way we have designed unit tests to mock better.

Reply:
And I think ^ would result in less work to change while getting the benefits of no longer using a global variable

Reply:
If you really prefer this being done in a follow up I won't stop this from going in, but this is something that I would think we want fixed very quickly. I don't think this is a good practice, even if it is in an internal package.

Reply:
well I'm part-way through with the swap out to interface, just doing the mocks needed in the tests now. If you are okay with it in a soon to be added PR, can we can get an approve on this PR and merge?